[AArch64] Custom lower v4i8 subreg extract.

davemgreen opened the pull request:
A v4i8 extract will usually be scalarized. This prevents that during lowering, converting it to an anyext and larger v4i16 subvector extract. There are a few minor regressions that are fixed up in the last patch in this commit.

The fold in DAGCombine was not handling the extended BUILDVECTORs that we see when i8/i16 are not legal types. Using isConstOrConstSplat(N1, false, true) allows it to match truncated constants. The other changes in visitAND are to make sure that truncated values in N1C are treated correctly, the fold we are mostly interested in is:
```
  if (N0.getOpcode() == ISD::EXTRACT_SUBVECTOR && N0.hasOneUse() && N1C &&
      ISD::isExtOpcode(N0.getOperand(0).getOpcode())) {
```
I couldn't find any other places where this triggered, as buildvectors are usually optimized before type legalization.

This also currently includes #133433.

efriedma-quic commented:
Can we just teach DAGTypeLegalizer::PromoteIntRes_EXTRACT_SUBVECTOR to do the right thing?

MacDue commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -7166,7 +7166,7 @@ SDValue DAGCombiner::visitAND(SDNode *N) {
   // if (and x, c) is known to be zero, return 0
   unsigned BitWidth = VT.getScalarSizeInBits();
-  ConstantSDNode *N1C = isConstOrConstSplat(N1);
+  ConstantSDNode *N1C = isConstOrConstSplat(N1, false, true);
```
nit: I find it helpful to label `bool` parameters.
```suggestion
  ConstantSDNode *N1C = isConstOrConstSplat(N1, /*AllowUndefs=*/false, /*AllowTruncation=*/true);
```

davemgreen committed:
[AArch64] Custom lower v4i8 subreg extract.

A v4i8 extract will usually be scalarized. This prevents that during lowering,
converting it to an anyext and larger v4i16 subvector extract. There are a few
minor regressions that are fixed up in a followup.

davemgreen commented:
> Can we just teach DAGTypeLegalizer::PromoteIntRes_EXTRACT_SUBVECTOR to do the right thing?

I gave it a go but I'm not sure how easy that would be. NVPTX and MVE both get stuck in a loop (MVE doesn't have a decent extend-half-vector-of operation, so needs an escape hatch to scalarize somewhere). There are the expected amount of regressions in all the X86 tests, and some of the other changes for AArch64 don't look too healthy. Some are improvements but I'm not sure if the others are easy to fix.

I was hoping to add v2i8 and v2i16 bitcast too, but those don't come up as much and v2 isn't a lot to scalarize.

davemgreen commented:
I made this into a SPR and moved the generic combine into #133915.

efriedma-quic commented:
> I gave it a go but I'm not sure how easy that would be

Okay, I won't push for it if it's too hard to make it work well.
